### PR TITLE
add snowflake destination

### PR DESCRIPTION
**Pull request by jrhizor**
Adds support for Snowflake as a destination.

## Recommended reading order
1. `snowflake.md`
1. `README.md` 
1. `SnowflakeDestination`
1. the rest

### CODE REVIEW

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeWriteConfig.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.destination.snowflake;
+
+import io.airbyte.commons.lang.CloseableQueue;
+
+class SnowflakeWriteConfig {
```
Comment (michel-tricot) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeWriteConfig.java:29]: NIT: SnowflakeWriteContext

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDestination.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    // we use transient instead of temporary because all of our statements don't run in a single session
+    for (final AirbyteStream stream : catalog.getStreams()) {
+      final String tableName = stream.getName();
+      final String tmpTableName = stream.getName() + "_" + Instant.now().toEpochMilli();
+
+      final String query = String.format(
+          "CREATE TRANSIENT TABLE IF NOT EXISTS \"%s\" ( \n"
+              + "\"ab_id\" VARCHAR PRIMARY KEY,\n"
+              + "\"%s\" VARIANT,\n"
+              + "\"emitted_at\" TIMESTAMP WITH TIME ZONE DEFAULT current_timestamp()\n"
```
Comment (michel-tricot) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDestination.java:116]: Shoud it be the current_timestamp or the one in the record?

**Review by michel-tricot — APPROVED** [review]
🎉

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDestination.java (hunk 2 above)
Comment (jrhizor) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDestination.java:116]: that's right, fixed.

File: docs/integrations/destinations/snowflake.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+```
+-- set variables (these need to be uppercase)
+set airbyte_role = 'AIRBYTE_ROLE';
+set airbyte_username = 'AIRBYTE_USER';
+set airbyte_warehouse = 'AIRBYTE_WAREHOUSE';
+set airbyte_database = 'AIRBYTE_DATABASE';
+set airbyte_schema = 'AIRBYTE_SCHEMA';
+
+-- set user password
+set airbyte_password = 'password';
```
Comment (cgardens) [inline docs/integrations/destinations/snowflake.md:44]: is it chill for password to be password?

Comment (cgardens) [reply docs/integrations/destinations/snowflake.md:44]: oh i guess it gets revoked afterwards?

File: airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/SnowflakeIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+public class SnowflakeIntegrationTest extends TestDestination {
+
+  private static final String COLUMN_NAME = "data";
+
+  // use a unique stream name from the source so we can write to the same schema/database in Snowflake
+  private static final String OLD_STREAM_NAME = "exchange_rate";
+  private static final String NEW_STREAM_NAME = "exchange_rate_" + RandomStringUtils.randomAlphanumeric(5);
+
+  @Override
+  protected Function<String, String> renameFunction() {
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-snowflake/src/test-integration/java/io/airbyte/integrations/destination/snowflake/SnowflakeIntegrationTest.java:50]: i think this is problematic. is this operation on the knowledge it knows that one of the streams that test runs internally includes `exchange_rate`? the implementation of this interface should not know the names of the streams (unless maybe they are provided through `testEnv`). we want to be able to add other datasets to test against, and we can't assume statically that their stream name will be `exchange_rate`

**Review by cgardens — CHANGES_REQUESTED** [review]
Looks good. Requesting changes because I want to understand what's going on with the renameFunction better. I think it might be breaking the contract of the standard tests.

File: docs/integrations/destinations/snowflake.md (hunk 3 above)
Comment (jrhizor) [reply docs/integrations/destinations/snowflake.md:44]: Above it says to change this to a real password. Should I add something here as well?

Comment (cgardens) [reply docs/integrations/destinations/snowflake.md:44]: sorry. i just didn't see it. i might do `<INSERT-YOUR-PASSWORD` instead `password`

**Review by cgardens — APPROVED** [review]
discussed offline with @jrhizor and agreed on a path forward that maintains the contract and gets what he needs.

**Comment by jrhizor** [comment]
Tested in the UI locally against exchangerate (just to sanity test) and against the past year of Stripe. Everything looks good. I'm going to resolve Charles's testing concern and then merge.

**Comment by michel-tricot** [comment]
Should we have snowflake in our end to end tests?

**Comment by cgardens** [comment]
> Should we have snowflake in our end to end tests?

@michel-tricot lol. this is underspecified which do you mean
1. Acceptance Tests
1. Integration tests that run for PRs
1. Integration tests that run for master (this is automatically the case)

**Comment by michel-tricot** [comment]
I meant acceptance tests :D

**Pull request merged**